src/main/java/com/google/firebase/auth/Tenant.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.firebase.auth;
+
+import com.google.auto.value.AutoValue;
+
+/**
+ * Contains metadata associated with a Firebase tenant.
+ * 
+ * <p>Instances of this class are immutable and thread safe.
+ */
+@AutoValue
+public abstract class Tenant {
+
+  /**
+   * Returns the ID of this tenant.
+   *
+   * @return a non-empty tenant ID string.
+   */
+  public abstract String getTenantId();
+
+  /**
+   * Returns the display name of this tenant.
+   *
+   * @return a non-empty display name string.
+   */
+  public abstract String getDisplayName();
+
+  /**
+   * Returns whether to allow email/password user authentication.
+   *
+   * @return true if a user can be authenticated using an email and password, and false otherwise.
+   */
+  public abstract boolean isPasswordSignUpAllowed();
+
+  /**
+   * Returns whether to enable email link user authentication.
+   *
+   * @return true if a user can be authenticated using an email link, and false otherwise.
+   */
+  public abstract boolean isEmailLinkSignInEnabled();
+
+  /**
+   * Returns a tenant builder.
+   */
+  public static Builder newBuilder() {
```

Comment:
I think we should make newBuilder() package-private. We don't expect clients to be able to use this builder directly, right?

Looks like I already had made the Builder class package private, but it's methods are public, which should be changed as well.

Reply:
Actually, do we ever need to build an instance of `Tenant`? It looks like they always get constructed from the JSON responses sent by the Auth backend service, in which case we can just rely on the JSON deserialization support we have.

Reply:
Are you suggesting that this class should look more like com.google.firebase.auth.UserRecord? And construct it with some sort of response and a JsonFactory? I'm not sure why this would mean we can't make it an AutoValue.

And I still think the inner CreateRequest and UpdateRequest classes will also make sense as AutoValues.

I don't think I have a good understanding of the API/JSON interaction yet. What is the best way to play around with the REST API and test sending various requests and getting reponses?

Reply:
We use Google API client to marshal/unmarshal JSON to Java types. If you have a `Tenant` class defined as follows:

```
class Tenant {
  @Key("tenantId")
  private String tenantId;
  
  @Key("displayName")
  private String displayName;
}
```

...then you can construct an instance of it from JSON as follows:

```
Tenant tenant = jsonFactory.createJsonParser(jsonString).parse(Tenant.class);
```

So AutoValue doesn't add anything to this flow as far as I can see. Create and Update classes are essentially builders, and I can see the value in using AutoValue for them.

Reply:
Alright, I've made the suggested changes.

I got the JSON key names from the Go API (https://github.com/firebase/firebase-admin-go/blob/e088f55036154f0cccf661992f854abd8fb1333f/auth/tenant_mgt.go).